### PR TITLE
Take two: Ensure `import urllib3` can succeed even if trio and twisted aren't installed

### PR DESCRIPTION
**Pull request by pquentin**
This pull request is based on @oliland's PR and completes it:

 * tests are now fixed
 * the demo is fixed
 * I've finally managed to understand [how the async test should work](https://github.com/python-trio/urllib3/pull/42#discussion_r186170793)

So the point of that test is simply to know whether we're running in async mode or bleached mode. We can only run in async mode with a Python that supports everything we need (3.6 right now), so the test is simple but has to be in a part of the code that is going to be bleached.

I decided to put this in _async/connection.py because it's the last place that accepts a backend, which means we can validate the backend in a single place. (Which also means that we did not really need to separate normalize_backend and load_backend.)

The history is getting a little complicated but the diff is not that big: I would recommend reviewers to look at the whole diff instead of individual commits.

### CODE REVIEW

File: urllib3/backends.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class Backend:
```
Comment (oliland) [inline urllib3/backends.py]: if this has to be py2 and 3 compatible, do we need to specify `(object)` here to guarantee new style classes?

File: urllib3/_backends/_loader.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+from ..backends import Backend
+
+
+class Loader:
```
Comment (oliland) [inline urllib3/_backends/_loader.py]: do we need to specify `(object)` here?

**Comment by pquentin** [comment]
@oliland Thanks for the review! You really did most of the work, including the import investigation: it was easy to add the finishing touches.

I think new-style classes are fine because we only support Python 2.7 which supports them. Did I miss something?

**Comment by pquentin** [comment]
Sorry, I did not leave you a lot of time to answer, but since you approved and since I think this PR does not contain new Python 2.7 issues, I'll go ahead and merge this. Thanks for your hard work, @oliland!

**Pull request merged**